lsp/ARCHITECTURE.md

```diff
@@ -0,0 +1,168 @@
+# Architechture
+
+## Language Service
+
+The following is a class diagram for the underlying Language Service.
+These are the components that enable users to provide a text document
+and get language support (hover, completion, ...) through Typescript code (as a Node module).
+
+This is separate from interfacing with it through a Language Server.
+
+### Example pseudocode:
+
+This is how 
+
+```typescript
+// Text document we want to do completion for
+incomingTextDocument: TextDocument = ...
+
+// Create document parser
+documentParserVisitor: DocumentParserVisitor = ...
+
+// Create language service strategy instances.
+// Note each strategy consumes the document parser, the strategy should know which parser to use.
+languageServiceStrategyA: LanguageServiceStrategy = new LanguageServiceStrategyA(documentParserVisitor)
+languageServiceStrategyB: LanguageServiceStrategy = new LanguageServiceStrategyB(documentParserVisitor)
+
+// Add strategies to the context
+languageServiceContext: LanguageServiceContext = ...
+languageServiceContenxt.addStrategy(languageServiceStrategyA)
+languageServiceContenxt.addStrategy(languageServiceStrategyB)
+
+// Resolve a specific Language Service instance from the context
+languageService: LanguageService = languageServiceContext.getLanguageService(incomingTextDocument)
+
+// Use the Language Service instance to do a completion
+completionResult = languageService.completion(incomingTextDocument, ...)
+```
+
+```mermaid
+classDiagram
+
+    class DocumentAST{
```

Comment:
Are these provided by an external node module or did we define and implement these? (if external, can we add a note?)

Reply:
We discussed this offline, the json and yaml types might become freestanding in the diagram (removing DocumentAst)

Reply:
Updated to add note explaining where the JSONDocument comes from (a separate node module)